open-craft/openedx-learning
# System defined taxonomies

## Description

Subclasses related to build System-defined taxonomies:

- [Docs updated](https://github.com/openedx/openedx-learning/pull/67/files#diff-e20a5299480027adca1a87aac4dea6d00797403cc0a23aab43bf0204ddb3a069) to reflect the new architecture of taxonomies.
- `models.py` moved to `models/base.py`
- `tag_ref` [setter](https://github.com/openedx/openedx-learning/pull/67/files#diff-7ffb9100f9426c81c59e88e9beceb9e3a8089da152d1d605b7de97990e6f0d50R540) to allow override how to set a tag on an object_tag
- `object_tag` [added to Taxonomy](https://github.com/openedx/openedx-learning/pull/67/files#diff-7ffb9100f9426c81c59e88e9beceb9e3a8089da152d1d605b7de97990e6f0d50R183)
- Base system-defined subclasses added.
- Language tags fixture.
- Taxonomy.system_defined updated to be a [hardcoded property](https://github.com/openedx/openedx-learning/pull/67/commits/1847248b48176014b989a25e03ec00b9dda5f26e#diff-7ffb9100f9426c81c59e88e9beceb9e3a8089da152d1d605b7de97990e6f0d50R203).
- TaxonomyAdmin updated to avoid edit and delete system taxonomies: https://github.com/openedx/openedx-learning/pull/67/commits/1847248b48176014b989a25e03ec00b9dda5f26e
- TagAdmin updated to avoid edit and delete tags from system taxonomies: https://github.com/openedx/openedx-learning/pull/67/commits/1847248b48176014b989a25e03ec00b9dda5f26e
- TagAmdin updated to avoid create Tags on system taxonomies: https://github.com/openedx/openedx-learning/pull/67/commits/1847248b48176014b989a25e03ec00b9dda5f26e

## Testing instructions

Manual testing is possible but not very useful, since we don't yet have views for these models and APIs.

But you can check that the added tests cover the expected polymorphism behavior supported by this change, and that tests are still covering 100% of the openedx_tagging module.

### Testing admin:

- On the backend shell create Taxonomies an Tags:

```
from openedx_tagging.core.tagging.models import Taxonomy, Tag
closed_taxonomy = Taxonomy(name='Test1', allow_free_text=False)
closed_taxonomy.save()
open_taxonomy = Taxonomy(name='Test2', allow_free_text=True)
open_taxonomy.save()
system_taxonomy = Taxonomy(name='Language', allow_free_text=False)
from openedx_tagging.core.tagging.models import LanguageTaxonomy
system_taxonomy.taxonomy_class = LanguageTaxonomy
system_taxonomy.save()
tag = Tag(taxonomy=closed_taxonomy, value='test')
tag.save()
tag = Tag(taxonomy=system_taxonomy, value='es')
tag.save()
```

- Go to [Django admin](http://127.0.0.1:8000/admin/oel_tagging/taxonomy/) and check that you can see all taxonomies.
- Click on the language taxonomy and check that you cannot edit or delete the taxonomy.
- Go to [Tags](http://127.0.0.1:8000/admin/oel_tagging/tag/) and check that you can see all tags.
- Click on the language tag and check that you cannot edit or delete the taxonomy.
- Create a Tag and check that you cannot add the tag to the open taxonomy nor system taxonomy

## Other information

Relates to https://github.com/openedx/modular-learning/issues/77
